### PR TITLE
Add a small documentation writer's guide

### LINKED ISSUE #7360: Developer documentation needed for the Sphinx manual
It would be good to have a developer documentation (`doc/sphinx/README.md`) explaining some basics about Sphinx, linking to the appropriate generic Sphinx documentation, explaining *all* Coq-Sphinx peculiarities and giving recommendation on how to extend / improve the Coq reference manual.

Of course, starting with something incomplete and improving it step by step would be OK.

Since @maximedenes is pretty busy ATM, @cpitclaudel would you be OK to take care of this? (I personally can't because I don't understand enough about Sphinx yet.)

Steps:

- [x] Create `doc/sphinx/README.md`.
- [x] Migrate relevant Emacs code to `dev/tools/coqdev.el`.
- [ ] Remove `doc/sphinx/MIGRATING`.

### PR DESCRIPTION
**Pull request by cpitclaudel**
**Kind:** documentation

Closes #7360.  You can see a rendered version at https://github.com/cpitclaudel/coq/tree/cpitclaudel_docs_devguide/doc/sphinx#documenting-coq-with-sphinx

### CODE REVIEW

File: doc/sphinx/README.template.rst
Comment (SkySkimmer) [inline doc/sphinx/README.template.rst]: Is the template really auto generated?

**Review by SkySkimmer — APPROVED** [review]
Fine for coqdev.el (without testing).

File: doc/sphinx/README.template.rst
Comment (cpitclaudel) [reply doc/sphinx/README.template.rst]: No, of course not :) — but what simpler way was there to get the comment in the README? I've just pushed a clarification. It now says this:

> README.rst is auto-generated from README.template.rst and the coqrst docs;
> use ``doc/tools/coqrst/regen_readme.py`` to rebuild it.

Comment (SkySkimmer) [reply doc/sphinx/README.template.rst]: >what simpler way was there to get the comment in the README

Generate it in the python program? Your solution is fine too.

File: doc/sphinx/README.rst
Comment (Zimmi48) [inline doc/sphinx/README.rst]: Either put trailing dots or provide the complete list... Isn't `@id` supposed to be `@ident` BTW?

File: doc/sphinx/README.rst
Comment (Zimmi48) [inline doc/sphinx/README.rst]: by commas

File: doc/sphinx/README.rst
Comment (Zimmi48) [inline doc/sphinx/README.rst]: Are you sure this shouldn't be:
```
:token:`term`
```
and
```
:token`ident`
```
instead?

File: doc/sphinx/README.rst
Comment (Zimmi48) [inline doc/sphinx/README.rst]: Typo: if

File: doc/sphinx/README.rst
Comment (Zimmi48) [inline doc/sphinx/README.rst]: The indentation here doesn't seem to be the usual one.

File: doc/sphinx/README.rst
Comment (Zimmi48) [inline doc/sphinx/README.rst]: This paragraph is interesting but either it is incomplete or I was given misleading information. Aren't there some automatic `Abort` also sent at the end of some `coqtop` directives? Or maybe it's only when leaving a block...

File: doc/sphinx/README.rst
Comment (Zimmi48) [inline doc/sphinx/README.rst]: Didn't know about this one. It seems that lots of places in the current manual are using `.. coqtop:: in` instead.

File: doc/sphinx/README.rst
Comment (Zimmi48) [inline doc/sphinx/README.rst]: Parenthesis is not closed. Should be transformed into a new sentence.

File: doc/sphinx/README.rst
Comment (Zimmi48) [reply doc/sphinx/README.rst]: And shouldn't it use
```
:token:`qualid`
```
?

File: doc/sphinx/README.rst
Comment (Zimmi48) [inline doc/sphinx/README.rst]: GitHub doesn't seem to understand this two line explanation. And note that verbatim should take two backquotes.

File: doc/sphinx/README.rst
Comment (Zimmi48) [inline doc/sphinx/README.rst]: Verbatim needs two backquotes.

**Review by Zimmi48 — CHANGES_REQUESTED** [review]
Thanks a lot! Here are a few remarks to address before merging this.

**Comment by Zimmi48** [comment]
Oh BTW, could you also remove the `MIGRATING` file as part of this PR? (Unless you still see value in it.)

File: doc/sphinx/README.rst
Comment (cpitclaudel) [reply doc/sphinx/README.rst]: That should look the same, though it's likely a bit slower.

File: doc/sphinx/README.rst
Comment (cpitclaudel) [reply doc/sphinx/README.rst]: Hmm. I don't think it ever calls abort, but there's a `reset` setting.

File: doc/sphinx/README.rst
Comment (cpitclaudel) [reply doc/sphinx/README.rst]: Yup, of course. I copied the examples directly from the manual, and I didn't pay enough attention to the actual contents :/ Fixed here and there.

File: doc/sphinx/README.rst
Comment (Zimmi48) [inline doc/sphinx/README.rst]: This internal reference does not seem to work. Maybe because you forgot to add a label...

File: doc/sphinx/README.rst
Comment (Zimmi48) [inline doc/sphinx/README.rst]: This link doesn't seem to work. Probably because you need to start it with `/` so that GitHub knows it is relative the the root of the repository... (Sorry, I have more experience with Markdown so I am not completely sure.)

File: doc/sphinx/README.rst
Comment (Zimmi48) [reply doc/sphinx/README.rst]: This looks like a comment you missed.

File: doc/sphinx/README.rst
Comment (Zimmi48) [inline doc/sphinx/README.rst]: Same problem with the links.
You can also list `|Ltac|` as part of the replacement patterns (it was introduced by @maximedenes when porting the Ltac chapter).
I also wondered why not all the chapters of the manual import the same two files. Should we enforce this?
And finally, is `\alors` a well-known LaTeX macro? To me it just sounds like a French word.

**Review by Zimmi48 — COMMENTED** [review]
Thanks!

File: doc/sphinx/README.rst
Comment (Zimmi48) [reply doc/sphinx/README.rst]: Thanks for clarifying this. Indeed, it has been falsely assumed by many people who did the migration work and because of Coq stupid nested lemmas support, it has gone largely unnoticed but we can find instances which demonstrate the problem: https://coq.inria.fr/refman/proof-engine/tactics.html#coq:tacn.decompose (subgoal remaining after `Qed`). Many a further warning in the documentation about abusing nested lemmas would be good.

File: doc/sphinx/README.rst
Comment (cpitclaudel) [reply doc/sphinx/README.rst]: Thanks

File: doc/sphinx/README.rst
Comment (cpitclaudel) [reply doc/sphinx/README.rst]: Correct

File: doc/sphinx/README.rst
Comment (cpitclaudel) [reply doc/sphinx/README.rst]: Yup,thanks

File: doc/sphinx/README.rst
Comment (cpitclaudel) [reply doc/sphinx/README.rst]: Done.

Comment (cpitclaudel) [reply doc/sphinx/README.rst]: > stupid nested lemmas support,

Well, I find it pretty convenient :smiley:

**Comment by cpitclaudel** [comment]
Done :) Thanks for the careful reviews

**Comment by Zimmi48** [comment]
> > I also wondered why not all the chapters of the manual import the same two files. Should we enforce this?
>
> We could, yes. Or we could introduce them lazily when they become needed.

Sure, but suppose one uses `|Coq|` in a chapter where it is not imported. Are they going to see a failure, or is it just going to look awful? Same question for other similar features that require importing one of these two files.

**Review by Zimmi48 — APPROVED** [review]
Looks very good. Thank you for preparing this!
@maximedenes I'll let you merge, unless you want me to do it.

**Comment by cpitclaudel** [comment]
> Sure, but suppose one uses |Coq| in a chapter where it is not imported. Are they going to see a failure, or is it just going to look awful?

They'll see an error, something like `Undefined substitution referenced: "Coq"`

**Comment by Zimmi48** [comment]
This PR apparently has some conflicts and needs to be rebased (@maximedenes don't forget to alert the author that an action on their part is wanted when you put this label).

**Review by maximedenes — APPROVED** [review]
Thanks a lot!

**Pull request merged**
